### PR TITLE
fix : Input control has dead-zones where clicking does not bring focus to the control

### LINKED ISSUE #30491: [Bug]: Input control has dead-zones where clicking does not bring focus to the control
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
System:
    OS: Linux 6.2 Ubuntu 22.04.3 LTS 22.04.3 LTS (Jammy Jellyfish)
    CPU: (32) x64 AMD EPYC 7763 64-Core Processor
    Memory: 80.74 GB / 125.78 GB
    Container: Yes
    Shell: 5.8.1 - /bin/zsh
```

### Are you reporting Accessibility issue?

no

### Reproduction

https://react.fluentui.dev/?path=/docs/components-input--default

### Bug Description

The Input control has areas surrounding the underlying input element, and clicking on these areas does not bring focus to the input control. This is unlike the native control, where the entire area of the control is clickable.

Repro:
Open https://react.fluentui.dev/?path=/docs/components-input--default
Use the mouse to click inside the Input control sample, very close to the border.

## Actual Behavior
Control does not receive focus when click happens near the borders (when mouse pointer is still an arrow).

## Expected Behavior
Control should receive focus when clicked anywhere inside its outer border.

## Additional information
The mouse pointer is a clue if you are hovering over an area that will bring focus to the input control or not - if it remains an arrow, it won't bring focus; if it switches to the text cursor, then we're actually on top of the underlying input element and clicking will bring focus to the control.

image

### Logs

_No response_

### Requested priority

Normal

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

yes

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by vicheanath**
fix: #30491

## Previous Behavior

The mouse pointer is a clue if you are hovering over an area that will bring focus to the input control or not - if it remains an arrow, it won't bring focus; if it switches to the text cursor, then we're actually on top of the underlying input element and clicking will bring focus to the control.

image

## New Behavior

Change root from `span` to `label` and put `:hover`

https://github.com/microsoft/fluentui/assets/48352653/64beab66-0f10-4b77-9131-61d55dc88e35

## Related Issue(s)

- Fixes #30491

### CODE REVIEW

**Comment by vicheanath** [comment]
@microsoft-github-policy-service agree

File: packages/react-components/react-input/src/components/Input/Input.types.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ export type InputSlots = {
    * The root slot receives the `className` and `style` specified directly on the `<Input>`.
    * All other top-level native props will be applied to the primary slot, `input`.
    */
-  root: NonNullable<Slot<'span'>>;
+  root: NonNullable<Slot<'label'>>;
```
Comment (spmonahan) [inline packages/react-components/react-input/src/components/Input/Input.types.ts:12]: I think this is a breaking API change. Putting a block on this PR while we consider the issue and how to handle it.

**Review by spmonahan — CHANGES_REQUESTED** [review]
Need to validate if the change is breaking and the correct thing to do.

File: packages/react-components/react-input/src/components/Input/Input.types.ts (hunk 1 above)
Comment (smhigley) [reply packages/react-components/react-input/src/components/Input/Input.types.ts:12]: I agree with @spmonahan, I don't think we can merge this change. Doing this would add the text in the contentBefore & contentAfter slots to the input's accessible name, which we almost certainly don't want.

The only way to override it would be via `aria-label` or `aria-labelledby`; using a `<label for="id">` wouldn't be enough. Since in v9 we want to support using built-in HTML accessibility features, I don't think we want to introduce a change that would force people to always use aria to label Inputs.

Comment (vicheanath) [reply packages/react-components/react-input/src/components/Input/Input.types.ts:12]: That is the first thing that I thought if if effects others. we can handle when clicking on span also focuses on input. if you have any idea I will follow. thanks

Comment (spmonahan) [reply packages/react-components/react-input/src/components/Input/Input.types.ts:12]: @vicheanath A better approach is to adjust the styles in `useInputStyle.styles.ts`. Something along the lines of this:

https://stackblitz.com/edit/stackblitz-starters-umxyha?description=React%20%20%20TypeScript%20starter%20project&file=src%2Findex.tsx,src%2FApp.tsx&title=React%20Starter

Also need to account for before/after slots.

Are you interested in making these changes?

Comment (vicheanath) [reply packages/react-components/react-input/src/components/Input/Input.types.ts:12]: i will

Comment (vicheanath) [reply packages/react-components/react-input/src/components/Input/Input.types.ts]: @spmonahan I have a complete change, if it still has any problem kindly feedback here.

File: packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts [hunk 2]
```diff
@@ -288,26 +344,30 @@ export const useInputStyles_unstable = (state: InputState): InputState => {
   state.input.className = mergeClasses(
     inputClassNames.input,
     useInputClassName(),
-    size === 'large' && inputStyles.large,
+    (size === 'large' && inputStyles.large) || (size === 'small' && inputStyles.small),
```
Comment (spmonahan) [inline packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts:347]: You can add an empty `medium` key to inputStyles (`medium: {}`) and then just index this on `size`

```suggestion
    inputStyles[size],
```

File: change/@fluentui-react-input-c8edf931-8ef0-4b0b-87e3-eff2133d203c.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
```
Comment (spmonahan) [inline change/@fluentui-react-input-c8edf931-8ef0-4b0b-87e3-eff2133d203c.json:2]:
```suggestion
  "type": "patch",
```

File: change/@fluentui-react-input-c8edf931-8ef0-4b0b-87e3-eff2133d203c.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+  "comment": "style : fix style input root on clickable",
+  "packageName": "@fluentui/react-input",
+  "email": "nathvichea1@gmail.com",
+  "dependentChangeType": "none"
```
Comment (spmonahan) [inline change/@fluentui-react-input-c8edf931-8ef0-4b0b-87e3-eff2133d203c.json:6]:
```suggestion
  "dependentChangeType": "path"
```

File: change/@fluentui-react-input-c8edf931-8ef0-4b0b-87e3-eff2133d203c.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "style : fix style input root on clickable",
```
Comment (spmonahan) [inline change/@fluentui-react-input-c8edf931-8ef0-4b0b-87e3-eff2133d203c.json:3]:
```suggestion
  "comment": "fix: update Input styles to improve clickability",
```

**Comment by vicheanath** [comment]
@spmonahan

**Comment by spmonahan** [comment]
> ```shell
> ERROR: there is an invalid dependentChangeType detected @fluentui-react-input-c8edf931-8ef0-4b0b-87e3-eff2133d203c.json: "path" is not a valid dependentChangeType
> ```
> 
> @spmonahan it seems like something wrong with dependentChangeType : "path", I'm not sure but it fails this path

Sorry! I looked and saw I told you to do the wrong thing. It should be `patch` with a "c" :)

File: packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts [hunk 6]
```diff
@@ -258,6 +259,58 @@ const useContentStyles = makeStyles({
...
+    paddingRight: 0,
+  },
+  medium: {
+    paddingRight: 0,
+  },
+  large: {
+    paddingRight: 0,
+  },
+});
+
```
[the remark this answers is not in the document]
Comment (vicheanath) [reply packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts:313]: that is a style that I want to apply on input when it has before or after. I need to check the condition when it has or not, I did not catch your idea.

[the remark this answers is not in the document]
Comment (vicheanath) [reply packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts:313]: or just change with your recommend

[the remark this answers is not in the document]
Comment (vicheanath) [reply packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts:313]: @sopranopillow  I'm not clear about your review can you explain me a bit more, I want to complete this PR

[the remark this answers is not in the document]
Comment (sopranopillow) [reply packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts:313]: Disregard the comment, I think I got the styles mixed up. Sorry for the confusion.

File: packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts [hunk 7]
```diff
@@ -28,7 +28,7 @@ const useRootClassName = makeResetStyles({
 
   // size: medium (default)
   minHeight: fieldHeights.medium,
-  padding: `0 ${tokens.spacingHorizontalMNudge}`,
+  padding: `0 ${tokens.spacingHorizontalXXS}`,
```
Comment (behowell) [inline packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts:31]: This will still leave some dead space at the start and end of the input. It's not a big deal, but since you're going to the effort here, might as well move all of the padding to the input slot when there's no before/after content.
```suggestion
  padding: '0',
```

Then, you can add the combined padding to the `input` slot when it doesn't have before/after content. And when there is before/after content, then adjust the padding on the `root` and `input` slots appropriately.  See my main review comment for the padding values to use.

File: packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts [hunk 8]
```diff
@@ -288,26 +345,30 @@ export const useInputStyles_unstable = (state: InputState): InputState => {
   state.input.className = mergeClasses(
     inputClassNames.input,
     useInputClassName(),
-    size === 'large' && inputStyles.large,
+    inputStyles[size],
```
Comment (behowell) [inline packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts:348]: Add the inputHasBeforeClassName, etc here instead of below inside the `if (state.contentBefore)`. The way you're adding them below will cause them to be overridden by `inputStyles[size]` since they'll be added ahead of those classes (later classes override earlier ones with atomic css).
```suggestion
    inputStyles[size],
    state.contentBefore && inputHasBeforeClassName[size],
    state.contentAfter && inputHasAfterClassName[size],
```

File: packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts [hunk 9]
```diff
@@ -258,6 +259,58 @@ const useContentStyles = makeStyles({
...
+  small: {
+    paddingRight: 0,
+  },
+  medium: {
+    paddingRight: 0,
+  },
+  large: {
+    paddingRight: 0,
+  },
+});
```
Comment (behowell) [inline packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts:312]: 0 padding doesn't match the previous padding values. Using the `padding` constants from my review comment, it should be:
```suggestion
const useInputHasBeforeClassName = makeStyles({
  small: {
    paddingLeft: padding.input.small,
  },
  medium: {
    paddingLeft: padding.input.medium,
  },
  large: {
    paddingLeft: padding.input.large,
  },
});
const useInputHasAfterClassName = makeStyles({
  small: {
    paddingRight: padding.input.small,
  },
  medium: {
    paddingRight: padding.input.medium,
  },
  large: {
    paddingRight: padding.input.large,
  },
});
```

File: packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts [hunk 10]
```diff
@@ -258,6 +259,58 @@ const useContentStyles = makeStyles({
...
+  medium: {
+    paddingRight: tokens.spacingHorizontalMNudge,
+  },
+  large: {
+    paddingRight: tokens.spacingHorizontalL,
+  },
+  disabled: {
+    color: tokens.colorNeutralForegroundDisabled,
+  },
+});
```
Comment (behowell) [inline packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts:290]: The contentBefore/contentAfter slots previously didn't have padding; it was on the root instead. Since it's possible for someone to customize the contentBefore/contentAfter slots to add padding, this would break those customizations. While we don't guarantee that styling changes won't be potentially breaking, we can reduce the likelihood of breaking something by keeping this padding on the root slot instead.

In short, change these to `useRootHasContentBeforeStyles`/`useRootHasContentAfterStyles`, and apply the padding to root instead.

**Review by behowell — COMMENTED** [review]
Thank you for the PR. I like these changes to move padding to the input slot. However, it's tricky to understand how all of the padding values add together, and relate them back to the design spec. I think that some of the padding values aren't correct in the PR as they are now.

It would be helpful to define all of the padding values as constants, which would make it easier to understand what's happening and verify that it matches the spec.

```ts
// If there is contentBefore or contentAfter, then the root and input slots use their individual padding.
// With no contentBefore or contentAfter, the padding is combined on the input slot to increase its hit target.
const padding = {
  root: {
    small: tokens.spacingHorizontalSNudge,
    medium: tokens.spacingHorizontalMNudge,
    large: tokens.spacingHorizontalM,
  },
  input: {
    small: tokens.spacingHorizontalXXS,
    medium: tokens.spacingHorizontalXXS,
    large: tokens.spacingHorizontalSNudge,
  },
  combined: {
    small: tokens.spacingHorizontalS, // SNudge + XXS
    medium: tokens.spacingHorizontalM, // MNudge + XXS
    large: `calc(${tokens.spacingHorizontalM} + ${tokens.spacingHorizontalSNudge})`,
  },
};
```

By default, `root` would have 0 padding, and `input` would use the combined padding. When the contentBefore or contentAfter slots are present, the padding would get applied to root or input appropriately.

I didn't leave comments in all of the places that would need to be updated, but basically anywhere that's setting left or right padding on the root or input would use the constants instead.

**Comment by behowell** [comment]
Hi @vicheanath, are you still working on this PR? If not, I'll make the changes I suggested in my review. Thanks!

**Comment by behowell** [comment]
> Hi @vicheanath, are you still working on this PR? If not, I'll make the changes I suggested in my review. Thanks!

I've created PR https://github.com/microsoft/fluentui/pull/30825 to implement the suggestions from my review.
